src/context.tsx

```diff
@@ -0,0 +1,53 @@
+import type { PropsWithChildren } from 'react';
+import * as React from 'react';
+import { createContext, useEffect, useState } from 'react';
+import type { To } from 'history';
+import { getLocationFromUrl, getReturnLocationFromSearch } from './utils';
+import { useLocation, useNavigate } from '@remix-run/react';
+
+interface ReturnNavigationContextData {
+  returnLocation: To | null;
+  defaultReturnLocationParam?: string;
+}
+
+type Props = PropsWithChildren<{
+  referrer: string | null;
+  defaultReturnLocationParam?: string;
+}>;
+
+export const ReturnNavigationContext = createContext<ReturnNavigationContextData>({
+  returnLocation: null,
+});
+
+export function ReturnNavigationContextProvider(props: Props) {
+  const [returnLocation, setReturnLocation] = useState(
+    props.referrer && getLocationFromUrl(props.referrer)
+  );
+
+  // Due to this issue, you cannot solely rely on the referer header
+  //   https://github.com/remix-run/remix/issues/3510
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    const returnLocation = getReturnLocationFromSearch(
+      location.search,
+      props.defaultReturnLocationParam
+    );
+
+    if (returnLocation) {
+      setReturnLocation(returnLocation);
+    }
+  }, [location, navigate, props.defaultReturnLocationParam, props.referrer]);
```

Comment:
Looks like you can kill navigate in this component

And if `props.defaultReturnLocationParam` and `props.referrer` are constantly changing, you can probs do more work outside the useeffect and keep the effect dependency real lean. It helps when craziness starts happening to easily reason about effects. 

```suggestion
    const returnLocation = getReturnLocationFromSearch(
      location.search,
      props.defaultReturnLocationParam
    );

  useEffect(() => {
    if (returnLocation) {
      setReturnLocation(returnLocation);
    }
  }, [returnLocation]);
```

Reply:
The referrer should be generated only the loader, so it should never change the value and the default param is also meant to be a statically passed config option, so it shouldn't change either.

Reply:
And, I killed the navigate.